clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+                            instr_t *instr) override;
+
+    instr_t *instr_ = nullptr;
+    void *dcontext_ = nullptr;
+};
+
+/**
+ * Base class for \p decode_cache_t.
+ *
+ * This is used to allow sharing the static data members among all template instances
+ * of \p decode_cache_t.
+ */
+class decode_cache_base_t {
+protected:
+    /**
+     * Constructor for the base class, intentionally declared as protected so
+     * \p decode_cache_base_t cannot be instantiated directly but only via
+     * a derived class.
+     */
+    decode_cache_base_t() = default;
+    virtual ~decode_cache_base_t();
+
+    static std::mutex module_mapper_mutex_;
+    static std::unique_ptr<module_mapper_t> module_mapper_;
+    // XXX: Maybe the ownership and destruction responsibility for the modfile bytes
+    // should be given to module_mapper_t instead.
+    static char *modfile_bytes_;
+    static int module_mapper_use_count_;
+
+    /**
+     * use_module_mapper_ describes whether we lookup the instr encodings
+     * from the module map, or alternatively from embedded-encodings in the
+     * trace.
+     *
+     * Note that we store our instr encoding lookup strategy as a non-static
+     * data member, unlike module_mapper_t which is static and shared between
+     * all decode_cache_t instances (even of different template types).
+     * Some analysis tools may deliberately want to look at instr encodings
+     * from the module mappings, but that strategy does not provide JIT
+     * encodings which are present only as embedded-encodings in the trace.
+     * In such a case, other concurrently running analysis tools should still
+     * be able to see encodings for JIT code.
+     */
+    bool use_module_mapper_ = 0;
+
+    /**
+     * Initializes the module_mapper_ object using make_module_mapper() and performs
+     * other bookkeeping and prerequisites.
+     */
+    std::string
+    init_module_mapper(const std::string &module_file_path,
+                       const std::string &alt_module_dir);
+
+    /**
+     * Returns the address where the encoding for the instruction at trace_pc can
+     * be found.
+     */
+    static std::string
+    find_mapped_trace_address(app_pc trace_pc, app_pc &decode_pc);
+
+private:
+    /**
+     * Creates a module_mapper_t. This does not need to worry about races as the
+     * module_mapper_mutex_ will be acquired before calling.
+     * Non-static to allow test sub-classes to override.
+     */
+    virtual std::string
+    make_module_mapper(const std::string &module_file_path,
+                       const std::string &alt_module_dir);
+};
+
+/**
+ * A cache to store decode info for instructions per observed app pc. The template arg
+ * DecodeInfo is a class derived from \p decode_info_base_t which implements the
+ * set_decode_info_derived() function that derives the required decode info from an
+ * \p instr_t object when invoked by \p decode_cache_t. This class handles the
```

Comment:
Should be # for a link; ditto below and elsewhere

Reply:
Done.